packages/apollo-link-utilities/src/index.ts

```diff
@@ -0,0 +1,269 @@
+     */
+    credentials?: string;
+
+    /**
+     * Any overrides of the fetch options argument to pass to the fetch call.
+     */
+    fetchOptions?: any;
+  }
+
+  const defaultHttpOptions: HttpOptions = {
+    includeQuery: true,
+    includeExtensions: false,
+  };
+
+  const defaultHeaders = {
+    // headers are case insensitive (https://stackoverflow.com/a/5259004)
+    accept: '*/*',
+    'content-type': 'application/json',
+  };
+
+  const defaultOptions = {
+    method: 'POST',
+  };
+
+  export const fallbackConfig = {
+    http: defaultHttpOptions,
+    headers: defaultHeaders,
+    options: defaultOptions,
+  };
+}
+
+export const throwServerError = (response, result, message) => {
+  const error = new Error(message) as ServerError;
+
+  error.response = response;
+  error.statusCode = response.status;
+  error.result = result;
+
+  throw error;
+};
+
+//TODO: when conditional types come in ts 2.8, operations should be a generic type that extends Operation | Array<Operation>
+export const parseAndCheckResponse = operations => (response: Response) => {
+  return (
+    response
+      .text()
+      .then(bodyText => {
+        try {
+          return JSON.parse(bodyText);
+        } catch (err) {
+          const parseError = err as ServerParseError;
+          parseError.response = response;
+          parseError.statusCode = response.status;
+          parseError.bodyText = bodyText;
+          return Promise.reject(parseError);
+        }
+      })
+      //TODO: when conditional types come out then result should be T extends Array ? Array<FetchResult> : FetchResult
+      .then((result: any) => {
+        if (response.status >= 300) {
+          //Network error
+          throwServerError(
+            response,
+            result,
+            `Response not successful: Received status code ${response.status}`,
+          );
+        }
+        //TODO should really error per response in a Batch based on properties
+        //    - could be done in a validation link
+        if (
+          !Array.isArray(result) &&
+          !result.hasOwnProperty('data') &&
+          !result.hasOwnProperty('errors')
+        ) {
+          //Data error
+          throwServerError(
+            response,
+            result,
+            `Server response was missing for query '${
+              Array.isArray(operations)
+                ? operations.map(op => op.operationName)
+                : operations.operationName
+            }'.`,
+          );
+        }
+        return result;
+      })
+  );
+};
+
+export const checkFetcher = (fetcher: GlobalFetch['fetch']) => {
+  warnIfNoFetch(fetcher);
+};
+
+const warnIfNoFetch = fetcher => {
+  if (!fetcher && typeof fetch === 'undefined') {
+    let library: string = 'unfetch';
+    if (typeof window === 'undefined') library = 'node-fetch';
+    throw new Error(
+      `fetch is not found globally and no fetcher passed, to fix pass a fetch for
```

Comment:
I think we should get rid of checking fetch and instead document that it needs to be polyfilled in readme setup instructions. This error message can't be minified and adds to the bundle size. The message recommending `node-fetch` might not make sense in a universal codebase: `isomorphic-unfetch` would be better.

Reply:
https://www.npmjs.com/package/invariant might be an option, if we retain dev errors.